### PR TITLE
Feat/naming specs

### LINKED ISSUE #57: Add start/end step utility for LTL specifications
Currently, SelVerifier automaticaly deduces the greatest but conservative start/end step. However, we should let users to specify the start/end step they want for convenience. Note that, if user specifies a greater (earlier for pLTL, later for fLTL) step than the automatically deduced step, SelVerifier should raise a specific error.

The semantics of `START` and `END` should be strict (we can change it to be non-strict as well), meaning that the verification for their own step should also be done, then the LTL verification ends with or without a success.

Example syntax:

```
...   { START phi }

.....

... { phi := <pLTL> }


//////////////////////////////////////////////7

... { psi := <fLTL> }

.....

.....{ END psi }

```

### LINKED ISSUE #69: (Optional) Fix Lark Parser Warnings
Current grammars `grammar.lark` and `spec_grammar.lark` both have several multiple definition warnings. It does not have any effect on the functionality at the moment, but can be easily fixed.

### PR DESCRIPTION
**Pull request by ykaydogdu**
Introduces the ability to give specs name in the SelVeri language. For example,
```
{ phi := Always x > 5 }
```
declares a future LTL spec with name `phi`. User can indicate custom ending points for the time domain of LTL statements using the syntax ```{ end phi }```. Similarly, start point can be defined for the past LTL. Seperated LTL statements can have both start and end markers.

Closes #57 

Furthermore, this PR also fixes all warnings in both `grammar.lark` and `spec_grammar.lark`, hence Closes #69 

The most notable change this PR introduces is that the parsing all specs before execution is removed. Now, verifier parses and processes all specs whenever they are sent by the interpreter at the `VERI` instruction.

### CODE REVIEW

**Comment by ykaydogdu** [comment]
Only failing test in CI is `fltl_named_end_before_spec_fail.svi` which throws a `CompilerError` as there is an invalid usage of end statement, but as CI script expects all `fail_tests` to throw `VerificationError`, the test fails.

CI behavior, or placement of that test, or even the decision of checking these in compile time must be discussed before merging this PR.

**Comment by iosephusferrum** [comment]
I have fixed the test script.

Also, I have added two fail tests ---about starting (ending) the verification earlier (later) than dependencies allow---  however the output was not expected. I have fixed the problem but I am not sure whether there are other checks we should be doing, for clean error detection. Please check my latest work.

Additionally, does this PR support using both `start` and `end` for sLTL formulae? If not, it must.

**Comment by ykaydogdu** [comment]
> Additionally, does this PR support using both `start` and `end` for sLTL formulae? If not, it must.

In the verification phase, as sLTL verification is implemented via pLTL and fLTL, there must not be any problems. We handle start markers regardless of the LTL type until it is encountered. End markers close `FutureObligation`s and one instance is generated for sLTL in the verification.

For compile time, there might be some edge cases, I will check and add some sLTL start end examples.

**Comment by iosephusferrum** [comment]
Can you please add sLTL examples too?

**Comment by ykaydogdu** [comment]
> Also, I have added two fail tests ---about starting (ending) the verification earlier (later) than dependencies allow--- however the output was not expected.

I have initially created `SpecDomainBoundsError` in `errors.py` for exactly this reason. However, while migrating to line by line processing of specs, I have followed a different approach. For **end**s, when exiting the scope, the `FutureObligation` is deleted and `{ end {name} }` refers to an obligation which does not exist anymore, so it naturally throws an error. For **start**s, variables used in the pLTL statement cannot be resolved, so ends up throwing a runtime error.

However, moving these checks into compile time makes sense. Thx for the commit

File: src/selveri/compiler.py [hunk 1]
```diff
@@ -187,29 +195,83 @@ def _finalize_spec_frame(self) -> None:
...
             self.emit("SPEC_START", name)
             return
-        if frame[name] is None:
+        if isinstance(frame[name], set) or frame[name] is None:
             raise CompilerError(f"Duplicate `{{ start {name} }}` in this scope.")
         raise CompilerError(
             f"`{{ start {name} }}` must appear before `{{ {name} := ... }}` in this scope."
         )
 
+    def _get_spec_free_vars(self, spec: Spec) -> set[str]:
```
Comment (ykaydogdu) [inline src/selveri/compiler.py:211]: Do we use this information to resolve the conservative start point for the pLTL statements so that we can throw an error if user uses `{ start phi }` before that point?

File: src/selveri/compiler.py [hunk 2]
```diff
@@ -187,29 +195,83 @@ def _finalize_spec_frame(self) -> None:
     def _emit_plt_start_marker(self, name: str) -> None:
         frame = self.spec_frames[-1]
         if name not in frame:
-            frame[name] = None
+            frame[name] = self.scope.get_all_bound_variables()
```
Comment (ykaydogdu) [inline src/selveri/compiler.py:202]: and I dont get why we insert information about bound variables at this point

**Comment by ykaydogdu** [comment]
Lastly, we might wait #74 to be merged before merging this as we need to forbid users from giving names to specs which are used as boolean expressions or obtain statements.

**Review by iosephusferrum — APPROVED** [review]
